### PR TITLE
copr: support pre-apply

### PR DESCRIPTION
**Pull request by BusyJay**
pre-apply is used to handle custom apply command.

### CODE REVIEW

File: src/raftstore/coprocessor/dispatcher.rs [hunk 1]
```diff
@@ -65,58 +65,37 @@ impl CoprocessorHost {
...
-        where H: FnMut(&mut RegionObserver, &mut ObserverContext, &Q, &mut R)
-    {
-        for entry in &mut self.registry.observers {
-            hook(entry.observer.as_mut(), &mut ctx, &req, resp);
-            if ctx.bypass {
-                break;
+    /// Call all pre apply hook until bypass is set to true.
+    pub fn pre_apply(&self, region: &Region, req: &mut RaftCmdRequest) {
+        let mut ctx = ObserverContext::new(region);
+        if !req.has_admin_request() {
```
Comment (siddontang) [inline src/raftstore/coprocessor/dispatcher.rs]: no need to handle admin request?

Comment (BusyJay) [inline src/raftstore/coprocessor/dispatcher.rs]: Not necessary yet.

**Comment by siddontang** [comment]
PTAL @hhkbp2 @zhangjinpeng1987

**Comment by siddontang** [comment]
@BusyJay 

You can help @lishihai9017 build a coprocessor to handle prewrite command.

File: src/raftstore/store/worker/apply.rs [hunk 2]
```diff
@@ -457,6 +460,7 @@ impl ApplyDelegate {
 
         let uuid = util::get_uuid_from_req(&cmd).unwrap();
         let cmd_cb = self.find_cb(uuid, term, &cmd);
+        apply_ctx.host.pre_apply(&self.region, &mut cmd);
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/apply.rs]: Is it means we should modify `cmd`? I think it is not friendly enough.

Comment (BusyJay) [inline src/raftstore/store/worker/apply.rs]: What do you recommend?

**Comment by zhangjinpeng1987** [comment]
LGTM

**Pull request merged**